sympy/tensor/tensor.py

```diff
@@ -3225,53 +3133,29 @@ def free_args(self):
     def components(self):
         return self._get_components_from_args(self.args)
 
-    @property
-    def free(self):
-        return self._index_structure.free[:]
-
     @property
     def free_in_args(self):
         arg_offset = self._get_position_offset_for_indices()
         argpos = self._get_indices_to_args_pos()
         return [(ind, pos-arg_offset[pos], argpos[pos]) for (ind, pos) in self.free]
 
-    @property
-    def coeff(self):
-        return self._coeff
-
     @property
     def nocoeff(self):
         return self.func(*[t for t in self.args if isinstance(t, TensExpr)]).doit()
 
-    @property
-    def dum(self):
-        return self._index_structure.dum[:]
-
     @property
     def dum_in_args(self):
         arg_offset = self._get_position_offset_for_indices()
         argpos = self._get_indices_to_args_pos()
         return [(p1-arg_offset[p1], p2-arg_offset[p2], argpos[p1], argpos[p2]) for p1, p2 in self.dum]
 
-    @property
-    def rank(self):
-        return len(self.free)
-
-    @property
-    def ext_rank(self):
-        return self._ext_rank
```

Comment:
The purpose of `@property` is to avoid assigning the value. You cannot do `t.ext_rank = ...`. Your edit is introducing this problem.

Reply:
Well, that is true. But there it is still possible to do `t._ext_rank = ...`, so what is the difference? And besides, as far as I understand, all objects in sympy are immutable so changing an object's property directly with an assignment is not a sympy way anyway. By the way, is there a sympy coding recommendations page somewhere, where such issues can be resolved once and for all?

Reply:
If the method name starts with an underscore `_`, it will be hidden from autocompletion. That is, when you use it in Jupyter/IPython you will not be prompted with `_ext_rank`. That's how you try to emulate private fields in Python... a bit poor compared to C++ or Java, but it works.